% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/findOverlaps.R
\name{overlapWithTx2}
\alias{overlapWithTx2}
\title{overlapWithTx2}
\usage{
overlapWithTx2(
  bamFile,
  annotation,
  uniqueFasta = NULL,
  ignoreStrand = TRUE,
  nbthreads = NULL
)
}
\arguments{
\item{bamFile}{A character vector of path to BAM file.}

\item{annotation}{An object of class `GRangesList`, as produced by
`prepareAnnotation()`.}

\item{uniqueFasta}{The path to the fasta file of unique fragments (which was
used for the alignment). This will is optional, but will be used to
recover the original sequence (rather than having arbitrary fragment IDs).}

\item{ignoreStrand}{Logical; whether to ignore strand when searching for
overlaps. By default, strand is ignored and considered later on at the read
assignment stage.}

\item{nbthreads}{A positive integer indicating the number of threads to use.
Defaults to `min(c(8, bpworkers()))`.}
}
\value{
A data.table.
}
\description{
Find overlaps between read alignments (bam format) and a feature annotation.
Supports spliced reads/transcripts, and calculates position relative to
transcript.
}
